### PR TITLE
Support Newtype validation in Scala 2 and 3

### PR DESCRIPTION
**Pull request by mvelimir**

### CODE REVIEW

File: modules/library/src/main/scala-2/zio/elasticsearch/IndexNameValidation.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+
+package zio.elasticsearch
+
+import org.apache.commons.lang3.StringUtils
+import org.apache.commons.lang3.StringUtils.{equalsAny, startsWithAny}
+
+object IndexNameValidation {
+  private def containsAny(string: String, params: List[String]): Boolean =
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-2/zio/elasticsearch/IndexNameValidation.scala:23]: Can we put this below public methods?

File: modules/library/src/main/scala-2/zio/elasticsearch/ValidatedNewtypes.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+package zio.elasticsearch
+
+import zio.prelude.Assertion.isEmptyString
+import zio.prelude.AssertionError.failure
+import zio.prelude.Newtype
+
+abstract class ValidatedNewtypes {
+  object IndexName extends Newtype[String] {
+    override def assertion = assertCustom { (string: String) => // scalafix:ok
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-2/zio/elasticsearch/ValidatedNewtypes.scala:25]: Maybe use `name` also.

File: modules/library/src/main/scala-3/zio/elasticsearch/IndexNameValidator.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+package zio.elasticsearch
+
+import org.apache.commons.lang3.StringUtils
+import org.apache.commons.lang3.StringUtils.{equalsAny, startsWithAny}
+import zio.prelude.{AssertionError, Validator}
+
+object IndexNameValidator
+    extends Validator[String](string => {
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-3/zio/elasticsearch/IndexNameValidator.scala:24]: I would use `name` here as well.

File: modules/library/src/main/scala-3/zio/elasticsearch/ValidatedNewtypes.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package zio.elasticsearch
+
+import zio.prelude.Assertion.isEmptyString
+import zio.prelude.{Newtype, NewtypeCustom}
+
+abstract class ValidatedNewtypes {
+  object IndexName extends NewtypeCustom[String] {
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-3/zio/elasticsearch/ValidatedNewtypes.scala:23]: Does it make sense to have separate objects for Index name and Routing. I would probably split this and remove `ValidatedNewtypes`.

File: modules/library/src/main/scala-3/zio/elasticsearch/IndexNameValidator.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        params.exists(StringUtils.contains(string, _))
+
+      def isValid(string: String): Boolean =
+        string.toLowerCase == string &&
+          !startsWithAny(string, "+", "-", "_") &&
+          !containsAny(string = string, params = List("*", "?", "\"", "<", ">", "|", " ", ",", "#", ":")) &&
+          !equalsAny(string, ".", "..") &&
+          string.getBytes().length <= 255
+
+      if (!isValid(string)) {
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-3/zio/elasticsearch/IndexNameValidator.scala:35]: Maybe we should use `if (isValid) Right(()) else ...`.

File: modules/library/src/main/scala-2/zio/elasticsearch/ValidatedNewtypes.scala [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      if (!IndexNameValidation.isValidIndexName(string)) {
```
Comment (drmarjanovic) [inline modules/library/src/main/scala-2/zio/elasticsearch/ValidatedNewtypes.scala:26]: Maybe to invert logic - `if (IndexNameValidation.isValidIndexName(...))`...

**Review by drmarjanovic — CHANGES_REQUESTED** [review]

**Review by arnoldlacko — DISMISSED** [review]

**Review by dbulaja98 — APPROVED** [review]

**Review by drmarjanovic — APPROVED** [review]

**Pull request merged**